File: src/main/kotlin/com/gildedrose/GildedRose.kt
package com.gildedrose

open class GildedItem(name: String, sellIn: Int, quality: Int): Item(name, sellIn, quality) {
    open fun age() {
        sellIn -= 1
    }
    open fun setQuality() {
        quality -= if (sellIn < 0) {2} else {1}
        if (quality < 0) {quality = 0}
    }
}

class AgedBrie(sellIn: Int, quality: Int): GildedItem("Aged Brie", sellIn, quality) {
    override fun setQuality() {
        quality += if (sellIn < 0) {2} else {1}
        if (quality > 50) {quality = 50}
    }
}
class BackstagePass(sellIn: Int, quality: Int): GildedItem("Backstage passes to a TAFKAL80ETC concert", sellIn, quality) {
    override fun setQuality() {
        quality += if (sellIn > 10) {1} else {if (sellIn > 5) {2} else {3}}
        quality = if (sellIn <= 0) {0} else {if (quality > 50) {50} else {quality}}
    }
}
class Sulfuras(sellIn: Int, quality: Int): GildedItem("Sulfuras, Hand of Ragnaros", sellIn, quality) {
    override fun age() {
        return
    }
    override fun setQuality() {
        return
    }
}
class ConjuredItem(name: String, sellIn: Int, quality: Int): GildedItem("Conjured $name", sellIn, quality) {
    override fun setQuality() {
        super.setQuality()
        super.setQuality()
    }
}

class GildedRose(var items: Array<GildedItem>) {

    fun updateQuality() {
        for (i in items.indices) {
            items[i].setQuality()
        }
        for (i in items.indices) {
            items[i].age()
        }
    }

}

File: app/gilded-rose.ts
export class Item {
    name: string;
    sellIn: number;
    quality: number;

    constructor(name, sellIn, quality) {
        this.name = name;
        this.sellIn = sellIn;
        this.quality = quality;
    }
}

export class GildedItem extends Item {
    constructor(name, sellIn, quality) {
        super(name, sellIn, quality);
    }
    age():void {
        this.sellIn -= 1
    }
    setQuality():void {
        if (this.sellIn <= 0) {this.quality -= 2} else {this.quality -= 1}
        this.checkQualityLimit()
    }
    checkQualityLimit():void {
        if (this.quality > 50) {this.quality = 50}
        if (this.quality < 0) {this.quality = 0}
    }
}
export class AgedBrie extends GildedItem {
    constructor(sellIn, quality) {
        super("Aged Brie", sellIn, quality);
    }
    setQuality():void {
        if (this.sellIn <= 0) {this.quality += 2} else {this.quality += 1}
        this.checkQualityLimit();
    }
}
export class BackstagePass extends GildedItem {
    constructor(sellIn, quality) {
        super("Backstage passes to a TAFKAL80ETC concert", sellIn, quality);
    }
    setQuality():void {
        if (this.sellIn > 10) {this.quality += 1} else {if (this.sellIn > 5) {this.quality += 2} else {this.quality += 3}}
        if (this.sellIn <= 0) {this.quality = 0}
        this.checkQualityLimit();
    }
}
export class Sulfuras extends GildedItem {
    constructor(sellIn, quality) {
        super("Sulfuras, Hand of Ragnaros", sellIn, quality);
    }
    age():void {
        return
    }
    setQuality():void {
        return
    }
}
export class ConjuredItem extends GildedItem {
    constructor(name, sellIn, quality) {
        super("Conjured " + name, sellIn, quality);
    }
    setQuality():void {
        super.setQuality()
        super.setQuality()
    }
}

export class GildedRose {
    items: Array<GildedItem>;

    constructor(items = [] as Array<GildedItem>) {
        this.items = items;
    }

    updateQuality() {
        for (var item of this.items) {
            item.setQuality();
            item.age();
        }
        return this.items;
    }
}
